cloud-exec.md

```diff
@@ -0,0 +1,153 @@
+   preserving the state.
+3. **Volumes.** Volumes can be attached and reused in instances to minimize
+   data cache synchronization time.
+4. **Shared volumes.** Separate cloud services (such as Multi-Attach EBS or
+   EFS) might be needed for sharing data cache between multiple instances.
+
+## Monitoring 
+
+Additional features might be needed to improve user experienc:
+1. **State monitoring.** A user should see if an instance was already
+   allocated, what stage is running.
+2. **Metrics monitoring.**  Latest metrics monitoring (with some reasonable
+   delay - 10-30-60 seconds)
+
+
+## Resource orchestrators
+
+1. AWS
+2. Azure
+3. GCP (optional)
+4. Remote over SSH
+5. Kubernetes (K8S)
+6. HashiCorp Nomad (optional)
+7. Container services (ECS and Azure/GCP analogs)
+
+
+# Detailed design
+
+## Iterative Terraform provider
+
+The cloud management part seems like a heavy part of the project (from
+  the implementation point of view). However, the problem of executing and
+  managing cloud instances was solved in the CML project.
+
+CML separates out the user API part (commands) and the cloud instace
+  management code part.
+The cloud management is extracted to a separate **Iterative Terraform (TF)
+  provider**:
+  https://github.com/iterative/terraform-provider-iterative
+
+The Iterative provider is specific to data scientists, not DevOps (TF's
+  target audience). It has a bit simplified API and some ML specific
+  functionality (some of these are coming) like: recover spot-instance if
+  it was deleted or attach a reusable drive.
+
+The Iteratve TF provider can be used from DVC for cloud instace management.
+Some Python libraries might be required  for access to GoLang specific
+terraform-provider.
+
+Today the provider supports:
+- AWS, Azure, K8S orchestrators
+- Spot instances and Volumes optimizations.
+A few of the initial scenarios can be already implemented with these
+functionalities. More functionality is comming.
+
+## Instances configuration
+
+Instance configuration is a definition of an instance and its behavior:
+- orchestrator: AWS, GCP, Azure, remote, K8S, ...
+- instance type
+- volume
+- many other options
+
+Instance configuration is well developed in CML project and can be reused.
+
+## Executors and naming
+
+Many instance types and even clouds could be used in a single DVC project.
+We need a way of configuring **executors** and reusing them the same way
+  as user can configure a data remote.
+It can look like this:
+```bash
+$ dvc exp run --executor my-gpu-tesla
+...
+$ dvc exp run --executor my-gpu-tesla8
+...
+$ dvc exp run --executor tesla8-ram
+```
+
+# How We Teach This
+
+We need to introduce a concept **Executor** to DVC.
+This convept should not break the compatibility and should not require
+significant changes in the docs (except command options).v
+However, it will require a new section in the docs to explain the cloud
+execution for the users who'd like to use this feature.
```

Comment:
If we are to abstract the execution environment, maybe we could start education from "local" executor - the machine user is using at the moment. I feel like explaining remotes starting from local ones made it easier for me.

Reply:
> start education from "local" executor

Good idea but that implies significant changes to docs (basically everywhere that different runs are mentioned we have to use the local executor concept). BTW I don't think we should avoid such big changes at all costs, just not sure it's necessary, may even make things unnecessarily complicated for people looking for a basic local workflow.

Reply:
The "local" executor already exists (`exp run --temp`), but we have already run into potential problems with making this a "default" behavior. Namely that debugging your pipeline becomes very difficult when the changes are not happening in your normal workspace.

So I think we would need to clarify what we want to do regarding educating the user about non-workspace execution environments.

Reply:
> The "local" executor already exists (exp run --temp)

Good point. Local (to the env) but external (to the project)? Like "local remotes".

Reply:
Actually I'm not sure. Maybe the local machine itself IS the executor, whether on a regular `repro` or an `exp run --temp`. Otherwise what exactly is an executor? 😅

Reply:
@pared could you please clarify? Do you mean abstracting environments like pip/conda packages or just directory (with `--temp` as @pmrowla mentioned)?

> Maybe the local machine itself IS the executor

@jorgeorpinel yes, it is the default executor. @pmrowla point is that we can abstract it out to a separate directory. But this is what `exp run --queue` does.

Reply:
@dmpetrov Not quite, I was rather thinking that we are already executing locally. In a way, our local environment is an executor, so what I though would be good is to teach this concept by saying "you already are using (local) executor - when using `--temp`". So as analogy - when you will be using remote executors, you will do similar thing as with `--temp` and local one - with all its pros and ceveats.